load-testing/initial-load-testing.md

```diff
@@ -0,0 +1,89 @@
+
+## Problems found during the tests
+
+### Timeouts
+There are 4 timeouts that were preventing large consignments being processed.
+* The akka request timeout. This is the maximum time a request can take
+* The akka server timeout. The maximum time a server response can take. This needs to be the same as the request timeout as they basically do the same thing.
+* The load balancer idle timout.
+* The client timeout when contacting the API. When the API is overloaded, requests can take longer and the clients are timing out.
+
+### Logging
+There is little to no logging in the backend checks lambdas. The antivirus prints the current file id but the others only log errors. This makes it difficult to find out what has happened when there are a lot of files processed. Also, we get one line of a stack trace per cloudwatch event which makes them difficult to read.
+
+### Client metadata api response json
+We are returning a large json payload from the client metadata mutation which is consuming a lot of memory in the browser and causing the browser to crash on lower spec machines.
+
+### No mulipart uploads in export task
+We are using the single file upload call in the export task which has a size limit of 5Gb
+
+### Parallelism in the backend check lambdas
+It seems that AWS is passing a lot of messages to individual lambdas which is making them slow. We can reduce the number of messages passed to an individual lambda but this will increase the cost as more lambdas will run.
+
+### Too many requests to the API
+The API update lambda is making a single request for every file which is partly why the API runs so slowly under load. 
+
+### API being overloaded
+The API was responding very slowly under load although the CPU and memory usage never moved much above 50%. This causes numerous problems inculding clients timing out and the web app itself running slowly for other users.
+
+### Lambda using all available memory
+The checksum and antivirus lambdas use all their available memory for larger files. Increasing it may improve the speed of the check but we'd have to run some more tests to be sure.
+
+## Potential fixes
+I'm going to group these roughly into quick fixes, medium effort fixes and more difficult changes so we can see what we want to prioritise.
+
+### Quick fixes
+* Add more logging to the lambdas and fix the multiline formatting so stack traces are printed correctly.
+* Change the graphql query used by the frontend to return less information.
+* Change the export task to use multipart uploads.
+
+### Medium effort fixes
+* Change the timeouts mentioned above. Changing the timeouts won't be difficult but testing to see what they should be will take time.
```

Comment:
I'd be quite cautious about significantly increasing the timeouts. Introducing more latency into a distributed system can have odd effects when the timeouts build up, and I don't think our monitoring is good enough to debug them yet!

Reply:
Agreed, see my comment above but I've moved it to the longer term fixes as we'll need to do quite a bit of investigation and the problem could be solved by other fixes.